scipy/ndimage/tests/test_interpolation.py

```diff
@@ -812,123 +938,143 @@ def test_affine_transform26(self, order):
         # see issue #4127
-        data = np.array([1])
+        data = xp.asarray([1])
         for out in [data.dtype, data.dtype.newbyteorder(),
-                    np.empty_like(data),
-                    np.empty_like(data).astype(data.dtype.newbyteorder())]:
-            returned = ndimage.affine_transform(data, [[1]], output=out)
+                    xp.empty_like(data),
+                    xp.empty_like(data).astype(data.dtype.newbyteorder())]:
+            returned = ndimage.affine_transform(data, xp.asarray([[1]]), output=out)
             result = out if returned is None else returned
-            assert_array_almost_equal(result, [1])
+            assert_array_almost_equal(result, xp.asarray([1]))
 
-    def test_affine_transform_output_shape(self):
+    def test_affine_transform_output_shape(self, xp):
         # don't require output_shape when out of a different size is given
-        data = np.arange(8, dtype=np.float64)
-        out = np.ones((16,))
+        if is_cupy(xp):
+            pytest.xfail("CuPy does not allow out of a different size.")
+
+        data = xp.arange(8, dtype=xp.float64)
+        out = xp.ones((16,))
 
-        ndimage.affine_transform(data, [[1]], output=out)
+        ndimage.affine_transform(data, xp.asarray([[1]]), output=out)
         assert_array_almost_equal(out[:8], data)
 
         # mismatched output shape raises an error
         with pytest.raises(RuntimeError):
             ndimage.affine_transform(
                 data, [[1]], output=out, output_shape=(12,))
 
-    def test_affine_transform_with_string_output(self):
-        data = np.array([1])
-        out = ndimage.affine_transform(data, [[1]], output='f')
-        assert_(out.dtype is np.dtype('f'))
-        assert_array_almost_equal(out, [1])
+    def test_affine_transform_with_string_output(self, xp):
+        data = xp.asarray([1])
+        out = ndimage.affine_transform(data, xp.asarray([[1]]), output='f')
+        assert out.dtype is np.dtype('f')
+        assert_array_almost_equal(out, xp.asarray([1]))
 
     @pytest.mark.parametrize('shift',
                              [(1, 0), (0, 1), (-1, 1), (3, -5), (2, 7)])
     @pytest.mark.parametrize('order', range(0, 6))
-    def test_affine_transform_shift_via_grid_wrap(self, shift, order):
+    def test_affine_transform_shift_via_grid_wrap(self, shift, order, xp):
         # For mode 'grid-wrap', integer shifts should match np.roll
-        x = np.array([[0, 1],
+        x = xp.asarray([[0, 1],
                       [2, 3]])
-        affine = np.zeros((2, 3))
-        affine[:2, :2] = np.eye(2)
-        affine[:, 2] = shift
+        affine = xp.zeros((2, 3))
+        affine[:2, :2] = xp.eye(2)
+        affine[:, 2] = xp.asarray(shift)
+
+        if is_cupy(xp):
+            expected = np.roll(x.get(), shift, axis=(0, 1))
+            expected = xp.asarray(expected)
+        else:
+            expected = np.roll(x, shift, axis=(0, 1))
```

Comment:
how about sticking with `np` until we get expected, then getting the `xp` versions of `x` and `expected`?

Reply:
@ev-br that isn't quite what I meant. We can avoid `is_cupy`:

```python
x = np.array([[0, 1],
              [2, 3]])
...
expected = xp.asarray(np.roll(x, shift, axis=(0, 1)))
x = xp.asarray(x)
```

Reply:
Yup, I'm having this in the follow-up, https://github.com/scipy/scipy/pull/21150. OK to defer to that one here?

Reply:
okay